### PR TITLE
Gatling JMS client concurrency issue

### PR DESCRIPTION
**Pull request by kiggundu**
The fix provided for issue #2913 in build 2.2.0-E20160217 worked fine agains the testcase provided in that ticket however when the number of concurrent users was increased the same unpredictable concurrency behaviour returned.
Looking at the fix it looks like the tracker was placed into the executing context which meant that a new one was created per context and thus per user. This pull request is a solution we coded which seems to be working fine for us under heavy user loads.

### CODE REVIEW

**Comment by slandelle** [comment]
> Looking at the fix it looks like the tracker was placed into the executing context which meant that a new one was created per context and thus per user.

Absolutely not. There's now only one single instance of `JmsRequestTrackerActor` per scenario. Message tracking then account for receiving queue.

No offense, but your proposed fix is wrong. It completely missed the whole point of Akka actors and what they guarantee in regard to the [JMM](http://doc.akka.io/docs/akka/2.4.2/general/jmm.html). There's absolutely no need for this `RequestResponseCorrelator` and all those Atomic and Concurrent structures, and all the more this scheduled `Runnable`.

For now, the only thing I see that could go wrong with thread-safety/memory visibility would be if your JMS implementation (which is, BTW?) was mutating the tracked `javax.jms.Message` after the `MessageProducer` sends them, or after the `MessageConsumer` receives them.

Please elaborate on your issue and the JMS implementation you're using, and even better, provide a reproducer.

**Comment by kiggundu** [comment]
Definitely no offense taken, I believe we all have vested interest in this being the best working solution possible. I must admit you're right, I have minimal experience with Akka much less its guarantees.

I have re-run the modified test-case that failed against your fix and completely failed to reproduce the failure (... no room for any more F's in that sentence).  I may well have not 'controlled' the test properly meaning the failure could have been caused by my queues not responding. Will ignore our changes at your advise for and keep you posted if that changes.

**Comment by slandelle** [comment]
> I have minimal experience with Akka much less its guarantees

TLDR;
Actor state is threadsafe as long as mutations only happen by message passing.

>  There's now only one single instance of JmsRequestTrackerActor per scenario

Correcting myself, `JmsRequestTrackerActor` is a proper global singleton. So using the same queue from multiple scenarios is safe.

> keep you posted if that changes

Please do. If you can get a reproducer, I'll gladly take a stab at it!

**Comment by kiggundu** [comment]
We continued to observe issues where though all messages were sent and replies picked up from the queues the internals of the JMS plugin sometimes tend to loose the replies. This is sporadic... i.e. sometimes it wrks and sometimes it doesn't so there is no real testcase i can submit. On looking again at the code I am still convinced that though there is only one tracker as per your fix and the advice above it is being hit concurrently by multiple threads yet it is not thread-safe wrt the sentMessages and receivedMessages fields.
 I produced the attached logs for the code at [this diff](https://github.com/gatling/gatling/compare/master...kiggundu:concurrencyFix?expand=1
)

[loadtest.usingtracker.log.txt](https://github.com/gatling/gatling/files/139770/loadtest.usingtracker.log.txt):
 This is with the code tracking the requests and responses within the JmsRequestTrackerActor as per the current implementation

[loadtest.externalsingleton.log.txt](https://github.com/gatling/gatling/files/139769/loadtest.externalsingleton.log.txt):
 This is with the code as per the stated diff (i.e. with the tracking variables synchronised. Only put them in there own singleton for clarity so don't think that's necessary)

because the sentmessages and receivedMessages fields are not protected you can see that the first logs are full of race conditions while the second are not. I would expect this to occur even if there was only one instance of the tracker unless i have greatly mis-understood the guarantees clarified above.

Please advise

**Comment by slandelle** [comment]
> because the sentmessages and receivedMessages fields are not protected you can see that the first logs are full of race conditions while the second are not.

Sorry, your test is broken if you're accessing the `sentmessages` and `receivedMessages` Actor  fields directly. Please check again the [Akka JMM guarantees](http://doc.akka.io/docs/akka/2.4.2/general/jmm.html): you must NEVER do that: mutations and memory visibility are only safe through message passing.

If you think you have an issue, please provide a demo app. But then, also consider the bug could be in your test or system under test. Are you sure that your correlationIds are random enough?

**Comment by slandelle** [comment]
PS: I'm turning `sentMessages` and `receivedMessages` visibility to private :)

**Comment by kiggundu** [comment]
That's a good move. 
We have definitely seen the behaviour again using the provided fix but on a more sporadic basis. Got 404's trying to access the fix's repository so had no option but to investigate using master but it could well be that our test has thus conflated both the issue that was fixed and another issue observed in the fixed code.

The "The actor subsequent processing rule" seems to be rule out my concurrency suspicions however, the fact is that we are still getting runs that appear to have internally lost replies consequently labeling them as errored/timed-out when it is obvious by looking at the queues that they were all sent back well before the timeout.

Can only really investigate further at our end if we have access to the fix repository though.

**Comment by slandelle** [comment]
Please get in touch at contact at gatling dot io
